# Assignment 1 House Prices

### Due: Wednesday, April 11th, 11:59 pm on Canvas.

In this assignment we'll train a linear regression model to predict house prices. Fill in the cells provided with code to answer the questions.

Answer the questions on Canvas and make sure to download this file once you've finished the assignment and upload it to Canvas as well.

Copyright ©2018 Emily Fox.  All rights reserved.  Permission is hereby granted to students registered for University of Washington CSE/STAT 416 for use solely during Spring Quarter 2018 for purposes of the course.  No other use, copying, distribution, or modification is permitted without prior written consent. Copyrights for third-party components of this work must be honored.  Instructors interested in reusing these course materials should contact the author.

In [102]:
import turicreate as tc # Fire up turicreate as usual and call it "tc" so we have to type less.

In [103]:
sales = tc.SFrame('/data/home_data.gl') # The dataset is stored in '/data/home_data.gl'.
sales


Columns:
	id	str
	date	datetime
	price	int
	bedrooms	int
	bathrooms	float
	sqft_living	int
	sqft_lot	int
	floors	float
	waterfront	int
	view	int
	condition	int
	grade	int
	sqft_above	int
	sqft_basement	int
	yr_built	int
	yr_renovated	int
	zipcode	str
	lat	float
	long	float
	sqft_living15	float
	sqft_lot15	float

Rows: 21613

Data:
+------------+---------------------------+---------+----------+-----------+-------------+
|     id     |            date           |  price  | bedrooms | bathrooms | sqft_living |
+------------+---------------------------+---------+----------+-----------+-------------+
| 7129300520 | 2014-10-13 00:00:00+00:00 |  221900 |    3     |    1.0    |     1180    |
| 6414100192 | 2014-12-09 00:00:00+00:00 |  538000 |    3     |    2.25   |     2570    |
| 5631500400 | 2015-02-25 00:00:00+00:00 |  180000 |    2     |    1.0    |     770     |
| 2487200875 | 2014-12-09 00:00:00+00:00 |  604000 |    4     |    3.0    |     1960    |
| 1954400510 | 2015-02-18 00:00:00+00:00 |  510000 |    3     |    2.0    |     1680    |
| 7237550310 | 2014-05-12 00:00:00+00:00 | 1225000 |    4     |    4.5    |     5420    |
| 1321400060 | 2014-06-27 00:00:00+00:00 |  257500 |    3     |    2.25   |     1715    |
| 2008000270 | 2015-01-15 00:00:00+00:00 |  291850 |    3     |    1.5    |     1060    |
| 2414600126 | 2015-04-15 00:00:00+00:00 |  229500 |    3     |    1.0    |     1780    |
| 3793500160 | 2015-03-12 00:00:00+00:00 |  323000 |    3     |    2.5    |     1890    |
+------------+---------------------------+---------+----------+-----------+-------------+
+----------+--------+------------+------+-----------+-------+------------+---------------+
| sqft_lot | floors | waterfront | view | condition | grade | sqft_above | sqft_basement |
+----------+--------+------------+------+-----------+-------+------------+---------------+
|   5650   |  1.0   |     0      |  0   |     3     |   7   |    1180    |       0       |
|   7242   |  2.0   |     0      |  0   |     3     |   7   |    2170    |      400      |
|  10000   |  1.0   |     0      |  0   |     3     |   6   |    770     |       0       |
|   5000   |  1.0   |     0      |  0   |     5     |   7   |    1050    |      910      |
|   8080   |  1.0   |     0      |  0   |     3     |   8   |    1680    |       0       |
|  101930  |  1.0   |     0      |  0   |     3     |   11  |    3890    |      1530     |
|   6819   |  2.0   |     0      |  0   |     3     |   7   |    1715    |       0       |
|   9711   |  1.0   |     0      |  0   |     3     |   7   |    1060    |       0       |
|   7470   |  1.0   |     0      |  0   |     3     |   7   |    1050    |      730      |
|   6560   |  2.0   |     0      |  0   |     3     |   7   |    1890    |       0       |
+----------+--------+------------+------+-----------+-------+------------+---------------+
+----------+--------------+---------+-------------+---------------+---------------+-----+
| yr_built | yr_renovated | zipcode |     lat     |      long     | sqft_living15 | ... |
+----------+--------------+---------+-------------+---------------+---------------+-----+
|   1955   |      0       |  98178  | 47.51123398 | -122.25677536 |     1340.0    | ... |
|   1951   |     1991     |  98125  | 47.72102274 |  -122.3188624 |     1690.0    | ... |
|   1933   |      0       |  98028  | 47.73792661 | -122.23319601 |     2720.0    | ... |
|   1965   |      0       |  98136  |   47.52082  | -122.39318505 |     1360.0    | ... |
|   1987   |      0       |  98074  | 47.61681228 | -122.04490059 |     1800.0    | ... |
|   2001   |      0       |  98053  | 47.65611835 | -122.00528655 |     4760.0    | ... |
|   1995   |      0       |  98003  | 47.30972002 | -122.32704857 |     2238.0    | ... |
|   1963   |      0       |  98198  | 47.40949984 | -122.31457273 |     1650.0    | ... |
|   1960   |      0       |  98146  | 47.51229381 | -122.33659507 |     1780.0    | ... |
|   2003   |      0       |  98038  | 47.36840673 |  

### Q1) How many houses do we have in our data set? Which column are we trying to predict given the other features (i.e. what's our output `y`)? How many inputs do we have in total in the data set (i.e. what is the size of our input `x`)?


Hint: You can use `len()` to get the length of anything in python. Note that when you use it on an SFrame, it will give you the number of rows, not columns.

a) 21613 houses, price, 20  
b) 21613 houses, bedrooms, 21  
c) 22014 houses, price, 21  
d) 22014 houses, bedrooms, 20  
e) 21613 houses, price, 21

In [104]:
len(sales)

21613

### Q2) What is the average price for the zipcode (neighborhood) with the most expensive house?

a) 75,000  
b) 7,700,000  
c) 540,088  
d) 2,160,607  
e) 901,258  

First find the zipcode with the highest listed house price.


In [105]:
sorted = sales.sort('price')
last_row = sorted[-1]
desired_zipcode = last_row.get('zipcode')
desired_zipcode




'98102'

Now find the average price of all the rows with that zipcode.  

In [106]:
all_houses_with_desired_zipcode = sales[sales['zipcode'] == desired_zipcode]
all_houses_with_desired_zipcode



Columns:
	id	str
	date	datetime
	price	int
	bedrooms	int
	bathrooms	float
	sqft_living	int
	sqft_lot	int
	floors	float
	waterfront	int
	view	int
	condition	int
	grade	int
	sqft_above	int
	sqft_basement	int
	yr_built	int
	yr_renovated	int
	zipcode	str
	lat	float
	long	float
	sqft_living15	float
	sqft_lot15	float

Rows: Unknown

Data:
+------------+---------------------------+---------+----------+-----------+-------------+
|     id     |            date           |  price  | bedrooms | bathrooms | sqft_living |
+------------+---------------------------+---------+----------+-----------+-------------+
| 1952200240 | 2014-06-11 00:00:00+00:00 |  850830 |    3     |    2.5    |     2070    |
| 1959700550 | 2014-09-05 00:00:00+00:00 |  740000 |    4     |    2.0    |     2050    |
| 1959701890 | 2014-07-29 00:00:00+00:00 |  865000 |    4     |    1.75   |     1800    |
| 6306400140 | 2014-06-12 00:00:00+00:00 | 1095000 |    0     |    0.0    |     3064    |
| 1962200435 | 2014-11-10 00:00:00+00:00 | 1010000 |    4     |    1.0    |     1820    |
| 6852700478 | 2014-09-16 00:00:00+00:00 |  425000 |    2     |    1.0    |     970     |
| 1962200036 | 2014-08-19 00:00:00+00:00 |  600000 |    3     |    1.75   |     1620    |
| 6003501535 | 2014-08-28 00:00:00+00:00 |  550000 |    3     |    1.75   |     1650    |
| 1336300445 | 2015-04-22 00:00:00+00:00 | 1265000 |    4     |    3.0    |     3130    |
| 3398800055 | 2014-11-19 00:00:00+00:00 | 2400000 |    4     |    3.75   |     4090    |
+------------+---------------------------+---------+----------+-----------+-------------+
+----------+--------+------------+------+-----------+-------+------------+---------------+
| sqft_lot | floors | waterfront | view | condition | grade | sqft_above | sqft_basement |
+----------+--------+------------+------+-----------+-------+------------+---------------+
|  13241   |  1.5   |     0      |  0   |     5     |   9   |    1270    |      800      |
|   4400   |  1.5   |     0      |  0   |     4     |   9   |    2050    |       0       |
|   4180   |  2.0   |     0      |  3   |     4     |   8   |    1800    |       0       |
|   4764   |  3.5   |     0      |  2   |     3     |   7   |    3064    |       0       |
|   5400   |  1.5   |     0      |  0   |     3     |   8   |    1820    |       0       |
|   2970   |  1.0   |     0      |  0   |     3     |   7   |    970     |       0       |
|   1325   |  2.5   |     0      |  0   |     3     |   9   |    1430    |      190      |
|   3200   |  1.5   |     0      |  0   |     3     |   8   |    1650    |       0       |
|   2646   |  2.5   |     0      |  0   |     5     |   9   |    2290    |      840      |
|  24825   |  2.0   |     0      |  0   |     4     |   11  |    3400    |      690      |
+----------+--------+------------+------+-----------+-------+------------+---------------+
+----------+--------------+---------+-------------+---------------+---------------+-----+
| yr_built | yr_renovated | zipcode |     lat     |      long     | sqft_living15 | ... |
+----------+--------------+---------+-------------+---------------+---------------+-----+
|   1910   |      0       |  98102  | 47.64152762 | -122.31541515 |     2200.0    | ... |
|   1922   |      0       |  98102  |  47.6439764 | -122.31867242 |     2320.0    | ... |
|   1921   |      0       |  98102  | 47.64623828 | -122.31803082 |     2180.0    | ... |
|   1990   |      0       |  98102  | 47.63624554 | -122.32156322 |     2360.0    | ... |
|   1923   |     2014     |  98102  | 47.64759138 | -122.31846469 |     1820.0    | ... |
|   1910   |      0       |  98102  | 47.62330469 | -122.31932474 |     1670.0    | ... |
|   2005   |      0       |  98102  | 47.64975713 |  -122.3209188 |     1750.0    | ... |
|   1901   |      0       |  98102  | 47.62091544 | -122.31734863 |     1500.0    | ... |
|   1906   |      0       |  98102  | 47.62722865 | -122.31617941 |     2920.0    | ... |
|   1926   |      0       |  98102  | 47.63384375 |

In [107]:
all_houses_with_desired_zipcode['price'].mean()

901258.238095238

### Q3) What fraction of properties are 2000-4000 sqft_living?

a) .2-.3
b) .3-.4
c) .4-.5
d) .5-.6
e) .6-.7

Find the number of rows where sqft_living >= 2000 and sqft_living < 4000, and divide it by the total number of rows in the table.

In [108]:
num_desired_houses = sales[(sales['sqft_living'] >=2000) & (sales['sqft_living'] < 4000)].num_rows()
num_desired_houses
num_desired_houses * 1.0 / sales.num_rows() 

0.4263174941007727

## Training a Linear Regression Model.

We will now train a linear regression model to make useful predictions. Work through the steps below and then answer the following questions.


Here we're splitting the data into a training and test set.

In [109]:
train_data, test_data = sales.random_split(.8, seed=0) # use seed=0

Now we're definining what our basic and advanced features we'll look at.

In [110]:
basic_features = ['bedrooms', 'bathrooms', 'sqft_living', 'sqft_lot', 'floors', 'zipcode']

In [111]:
advanced_features = ['condition', # condition of house
                     'grade', # measure of qality of construction
                     'waterfront', # waterfront property 
                     'view', # type of view
                     'sqft_above', # square feet above ground
                     'sqft_basement', # square feet in basement
                     'yr_built', # the year built
                     'yr_renovated', # the year renovated
                     'lat', 'long', # the lat-long of the parcel
                     'sqft_living15', # average sq.ft. of 15 nearest neighbors 
                     'sqft_lot15', # average lot size of 15 nearest neighbors 
                    ]

In [112]:
all_features = basic_features + advanced_features

Uncomment the following lines and fill in the parameters to create and train two linear regression models. One model should use only the basic features, and the other model should use all the features.

Only fill in the parameters for `dataset`, `target` and `features`.

Look at the API for Turi Create's Linear Regression model here to understand how to use it:
https://apple.github.io/turicreate/docs/api/generated/turicreate.linear_regression.LinearRegression.html

Now, evaluate the models on the test set and print the results.

### Q4) What are your RMSE errors on your training data using the basic model and all features model?

a) ~ 4090457, 4378167  
b) ~ 189236, 162526    
c) ~ 3157653, 3159982  
d) ~ 180498, 154573 

In [142]:
basic_model = tc.linear_regression.create(train_data, target='price', features=basic_features)
all_features_model = tc.linear_regression.create(train_data, target='price', features=all_features)

PROGRESS: Creating a validation set from 5 percent of training data. This may take a while.
          You can set ``validation_set=None`` to disable validation tracking.



Linear regression:

--------------------------------------------------------

Number of examples          : 16536

Number of features          : 6

Number of unpacked features : 6

Number of coefficients    : 75

Starting Newton Method

--------------------------------------------------------

+-----------+----------+--------------+--------------------+----------------------+---------------+-----------------+

| Iteration | Passes   | Elapsed Time | Training-max_error | Validation-max_error | Training-rmse | Validation-rmse |

+-----------+----------+--------------+--------------------+----------------------+---------------+-----------------+

| 1         | 2        | 0.111736     | 4086563.803876     | 1885934.832130       | 189713.101374 | 179517.689735   |

+-----------+----------+--------------+--------------------+----------------------+---------------+-----------------+

PROGRESS: Creating a validation set from 5 percent of training data. This may take a while.
          You can set ``validation_set=None`` to disable validation tracking.



SUCCESS: Optimal solution found.

Linear regression:

--------------------------------------------------------

Number of examples          : 16524

Number of features          : 18

Number of unpacked features : 18

Number of coefficients    : 87

Starting Newton Method

--------------------------------------------------------

+-----------+----------+--------------+--------------------+----------------------+---------------+-----------------+

| Iteration | Passes   | Elapsed Time | Training-max_error | Validation-max_error | Training-rmse | Validation-rmse |

+-----------+----------+--------------+--------------------+----------------------+---------------+-----------------+

| 1         | 2        | 0.030821     | 4332617.258386     | 1377360.304727       | 161841.616173 | 173331.193831   |

+-----------+----------+--------------+--------------------+----------------------+---------------+-----------------+

SUCCESS: Optimal solution found.

### Q5) What are your RMSE errors on your test data using the basic model and all features model?


a) ~ 4090457, 4378167  
b) ~ 189236, 162526    
c) ~ 3157653, 3159982  
d) ~ 180498, 154573  

In [143]:
basic_model = tc.linear_regression.create(test_data, target='price', features=basic_features)
all_features_model = tc.linear_regression.create(test_data, target='price', features=all_features)

PROGRESS: Creating a validation set from 5 percent of training data. This may take a while.
          You can set ``validation_set=None`` to disable validation tracking.



Linear regression:

--------------------------------------------------------

Number of examples          : 4026

Number of features          : 6

Number of unpacked features : 6

Number of coefficients    : 75

Starting Newton Method

--------------------------------------------------------

+-----------+----------+--------------+--------------------+----------------------+---------------+-----------------+

| Iteration | Passes   | Elapsed Time | Training-max_error | Validation-max_error | Training-rmse | Validation-rmse |

+-----------+----------+--------------+--------------------+----------------------+---------------+-----------------+

| 1         | 2        | 0.030899     | 2551220.812341     | 548957.146738        | 176451.256057 | 139467.791460   |

PROGRESS: Creating a validation set from 5 percent of training data. This may take a while.
          You can set ``validation_set=None`` to disable validation tracking.



+-----------+----------+--------------+--------------------+----------------------+---------------+-----------------+

SUCCESS: Optimal solution found.

Linear regression:

--------------------------------------------------------

Number of examples          : 4013

Number of features          : 18

Number of unpacked features : 18

Number of coefficients    : 87

Starting Newton Method

--------------------------------------------------------

+-----------+----------+--------------+--------------------+----------------------+---------------+-----------------+

| Iteration | Passes   | Elapsed Time | Training-max_error | Validation-max_error | Training-rmse | Validation-rmse |

+-----------+----------+--------------+--------------------+----------------------+---------------+-----------------+

| 1         | 2        | 0.130961     | 2430783.532718     | 801949.929382        | 147859.400187 | 160493.791105   |

+-----------+----------+--------------+--------------------+----------------------+---------------+-----------------+

SUCCESS: Optimal solution found.

### Q6) a. Which model would you choose?

a) Model with the basic features  
b) Model with the all the features


### b. Why?

a) It has higher training error  
b) It uses more features  
c) It has lower test error  
d) It has lower training error  
e) It has higher test error

### Q7) Using the basic features model, predict the house prices for the houses with ids 1336300445 and 2724200705. What are their predicted prices?

Remember, you can index into the Sframe to find a particular id like this `sales[sales['id'] == ?]`.

a) ~ \$1265425, \$95000 respectively  
b) ~ \$1265000, \$95000 respectively  
c) ~ \$2200080, \$191428 respectively   
d) ~ \$1166712, \$93198 respectively

In [152]:
price1 = sales[sales['id'] == '1336300445']
price1_basic_features = price1['bedrooms', 'bathrooms', 'sqft_living', 'sqft_lot', 'floors', 'zipcode']
predictions1 = basic_model.predict(price1_basic_features)
print(predictions1)


[1051684.996949727]


In [153]:
price2 = sales[sales['id'] == '2724200705']
price2_basic_features = price2['bedrooms', 'bathrooms', 'sqft_living', 'sqft_lot', 'floors', 'zipcode']

predictions2 = basic_model.predict(price2_basic_features)
print(predictions2)

[96244.2220178077]
